File: frontend-admin/src/api/monitor.ts
import { api } from "./client";
import type { RosterSitting } from "./types";

export interface SessionSummary {
  session_id: string;
  candidate_id: string;
  cccd: string;
  full_name: string;
  unit: string;
  category: string;
  attempt_number: number;
  photo_path: string | null;
  status: string;
  submitted_at: string | null;
  /** Mốc hết giờ RIÊNG của thí sinh này (đồng hồ là per-candidate, AD-47). */
  end_time: string | null;
  paused: boolean;
  /** Số khiếu nại câu hỏi CHƯA xử lý của thí sinh này. */
  open_question_reports: number;
  /** Lý do bị đình chỉ thi (status = 'terminated'). */
  terminated_reason: string | null;
  /** Đang tạm dừng VÀ đã quá end_time — bài sẽ không bao giờ tự nộp (AD-121 #2). */
  overdue_paused: boolean;
  self_registered: boolean;
  room_id: string | null;
  room_name: string | null;
  // AD-110: máy đã tải xong toàn bộ ảnh đề (báo về từ máy thí sinh lúc chờ).
  preloaded: boolean;
  /** Thí sinh đã bấm "Báo giám thị" (sai thông tin), chưa được sửa. */
  info_disputed: boolean;
  /** Máy im lặng >90s = mất kết nối THẬT (chớp mạng ngắn không tính). */
  offline: boolean;
  last_seen_seconds: number | null;
}

export interface RosterCandidate {
  candidate_id: string;
  cccd: string;
  full_name: string;
  unit: string;
  category: string;
  attempt_number: number;
  photo_path: string | null;
  self_registered: boolean;
  room_name: string | null;
  info_disputed: boolean;
}

export interface RosterResponse {
  sitting: RosterSitting;
  assigned_total: number;
  logged_in: number;
  not_logged_in_total: number;
  not_logged_in: RosterCandidate[];
  // Đồng hồ thi chung (AD-78): deadline sớm nhất trong các phiên đang làm + giờ server.
  earliest_end_time: string | null;
  /** Mốc của người kết thúc muộn nhất (vào trễ / cộng giờ riêng). */
  latest_end_time: string | null;
  server_time: string | null;
}

/** Session-level controls (chủ tịch on any owned exam; giám thị on their room). */
export const monitorApi = {
  pauseSession: async (sessionId: string) =>
    (await api.post(`/admin/sessions/${sessionId}/pause`)).data,
  resumeSession: async (sessionId: string) =>
    (await api.post(`/admin/sessions/${sessionId}/resume`)).data,
  /** Thoát phần mềm thi trên đúng MÁY của một thí sinh (AD-128). */
  kioskQuitSession: async (sessionId: string): Promise<{ targeted: boolean; detail?: string }> =>
    (await api.post(`/admin/sessions/${sessionId}/kiosk-quit`)).data,
  logout: async (sessionId: string) =>
    (await api.post(`/admin/sessions/${sessionId}/logout`)).data,
  admit: async (sessionId: string) =>
    (await api.post(`/admin/sessions/${sessionId}/admit`)).data,
  /** Cộng giờ cho RIÊNG một thí sinh (máy treo/hỏng) — không đụng cả phòng. */
  extendSession: async (sessionId: string, minutes: number) =>
    (await api.post(`/admin/sessions/${sessionId}/extend`, { minutes })).data,
  /** Đình chỉ thi: dừng hẳn bài, chấm với những gì đã làm. Lý do bắt buộc. */
  terminateSession: async (sessionId: string, reason: string) =>
    (await api.post(`/admin/sessions/${sessionId}/terminate`, { reason })).data,
  /** Khiếu nại câu hỏi của thí sinh trong buổi này (chưa xử lý lên trước). */
  questionReports: async (sittingId: string): Promise<QuestionReport[]> =>
    (await api.get(`/admin/sittings/${sittingId}/question-reports`)).data,
  resolveQuestionReport: async (reportId: string, resolution: string) =>
    (await api.post(`/admin/question-reports/${reportId}/resolve`, { resolution })).data,
};

export interface QuestionReport {
  id: string;
  candidate_id: string;
  cccd: string;
  full_name: string;
  room_name: string | null;
  question_id: string;
  /** Số thứ tự câu TRONG ĐỀ CỦA THÍ SINH ĐÓ (đề trộn nên mỗi người một thứ tự). */
  question_number: number;
  content: string;
  created_at: string;
  resolved_at: string | null;
  resolution: string | null;
  resolved_by_name: string | null;
}

File: backend/app/schemas/monitor.py
"""Schemas for exam control + monitoring + security report."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ExtendRequest(BaseModel):
    minutes: int = Field(ge=1, le=180)


class StartResult(BaseModel):
    started: int
    end_time: datetime | None


class EndResult(BaseModel):
    submitted: int


class SessionSummary(BaseModel):
    session_id: uuid.UUID
    candidate_id: uuid.UUID
    cccd: str
    full_name: str
    unit: str
    category: str
    attempt_number: int
    photo_path: str | None
    status: str
    started_at: datetime | None
    submitted_at: datetime | None
    end_time: datetime | None = None
    score: float | None
    total_correct: int | None
    client_ip: str | None
    device_id: str | None = None
    self_registered: bool = False
    # Per-candidate pause + room assignment (AD-47).
    paused: bool = False
    # Đang tạm dừng VÀ đồng hồ đã trôi qua end_time. Vòng quét tự nộp cố ý bỏ qua
    # phiên tạm dừng (không nộp thay người đang bị dừng), nên nếu không ai bấm Tiếp
    # tục thì phiên nằm đó mãi. Cờ này để bảng giám sát làm nó nổi lên thay vì im
    # lặng treo (lỗ AD-121 #2).
    overdue_paused: bool = False
    room_id: uuid.UUID | None = None
    room_name: str | None = None
    # AD-122: thí sinh đã bấm "Báo giám thị" (sai thông tin) và CHƯA được sửa.
    info_disputed: bool = False
    # AD-122: máy thí sinh im lặng quá OFFLINE_ALERT_SECONDS (90s) — mất kết nối
    # THẬT chứ không phải chớp mạng. Chỉ tính khi phiên còn đang cần online.
    offline: bool = False
    last_seen_seconds: int | None = None
    # AD-110: máy đã tải xong toàn bộ ảnh đề (cờ Redis do máy thí sinh báo về) —
    # chủ tịch chỉ nên Bắt đầu thi khi mọi máy ready đều True.
    preloaded: bool = False


class RosterCandidate(BaseModel):
    candidate_id: uuid.UUID
    cccd: str
    full_name: str
    unit: str
    category: str
    attempt_number: int
    photo_path: str | None
    self_registered: bool = False
    room_name: str | None = None
    info_disputed: bool = False


class RosterSitting(BaseModel):
    """Header for the monitor — describes the SITTING being watched (AD-47)."""
    sitting_id: uuid.UUID
    exam_id: uuid.UUID
    exam_name: str
    sitting_name: str
    exam_date: str | None
    duration_minutes: int
    status: str
    question_count: int


class RosterResponse(BaseModel):
    sitting: RosterSitting
    assigned_total: int
    logged_in: int
    not_logged_in_total: int
    not_logged_in: list[RosterCandidate]
    # Mốc đếm ngược ở màn giám sát (AD-78). Đồng hồ là per-candidate nên không có
    # deadline chung: ``earliest`` = người xong sớm nhất, ``latest`` = người xong
    # muộn nhất (vào trễ / được cộng giờ riêng). Cả hai đều BỎ QUA phiên đang tạm
    # dừng — đồng hồ của họ đóng băng nên end_time không phản ánh thời gian thực
    # của ai cả (lỗ AD-121 #3). server_time để máy admin bù lệch đồng hồ.
    earliest_end_time: datetime | None = None
    latest_end_time: datetime | None = None
    server_time: datetime | None = None


class SecurityEventOut(BaseModel):
    id: uuid.UUID
    event_type: str
    cccd_attempted: str | None
    client_ip: str | None
    created_at: datetime
    metadata: dict | None
